cli/handleSysCheck.js

```diff
@@ -13,13 +23,115 @@ const renderHelp = () => `
     $ yeep syscheck --config=yeep.config.js
 `;
 
+const validateMongo = async (uri, spinner) => {
+  const client = await MongoClient.connect(
+    uri,
+    { useNewUrlParser: true }
+  );
+
+  spinner.stopAndPersist({
+    symbol: chalk.green('✔'),
+    text: 'Connected to database',
+  });
+
+  const db = client.db();
+  try {
+    // TODO: Verify that the uri has the username:password format if the authentication fails!
+    const buildInfo = await db.admin().buildInfo();
+    const isVersionValid = buildInfo.versionArray[0] >= 4;
+    if (!isVersionValid) {
+      spinner.fail(`Mongo version ${buildInfo.version} needs to be higher than 4.0.0`);
+      throw new Error('Mongo version missmatch');
+    }
+    spinner.stopAndPersist({
+      symbol: chalk.green('✔'),
+      text: `Mongo version ${buildInfo.version} is valid (>= 4.0.0)`,
+    });
+
+    let replStatus;
+    try {
+      replStatus = await db.admin().replSetGetStatus();
+    } catch (err) {
+      spinner.fail(`Mongo database needs to exist within a replica set`);
+      throw err;
+    }
+    spinner.stopAndPersist({
+      symbol: chalk.green('✔'),
+      text: `Mongo replica set ${replStatus.set} found`,
+    });
+
+    // check migrations
+    const dbMigrator = new DatabaseMigrator({
+      mongoUri: uri,
+      migrationDir: path.resolve(__dirname, '../migrations'),
+    });
+
+    await dbMigrator.connect();
+    const pendingMigrations = await dbMigrator.findPendingMigrations();
+    await dbMigrator.disconnect();
+    if (!pendingMigrations || pendingMigrations.length !== 0) {
+      throw new Error('Migrations have not been applied successfully');
+    }
+  } catch (err) {
+    throw err;
+  } finally {
+    await client.close();
+  }
+};
+
+const validateStorage = async (type, directory, spinner) => {
+  if (type !== 'fs') {
+    return;
+  }
+  const dirExists = await dirExistsAsync(directory);
+  if (!dirExists) {
+    throw new Error(`Directory ${directory} does not exist`);
+  }
+  spinner.stopAndPersist({
+    symbol: chalk.green('✔'),
+    text: 'Upload directory exists',
+  });
+
+  try {
+    await accessAsync(directory, fs.W_OK);
+  } catch (err) {
+    spinner.fail(`Directory ${directory} has no write permissions`);
+    throw err;
+  }
+
+  spinner.stopAndPersist({
+    symbol: chalk.green('✔'),
+    text: 'Upload Directory has write permissions',
+  });
+};
+
 const handleSysCheck = (inputArr, flagsObj) => {
   if (flagsObj.help) {
     console.log(renderHelp());
   } else if (!flagsObj.config) {
     console.error(renderMissingConfig());
   } else {
-    console.log(123);
+    let config;
+    const configPath = path.resolve(flagsObj.config);
+    try {
+      config = require(configPath);
+    } catch (err) {
+      console.error(renderNativeError(err));
+    }
+    const spinner = ora();
+    spinner.start('Checking system status...');
+
+    Promise.all([
```

Comment:
I see what you are doing here, but I think these 2 methods should run in series, instead of parallel...

```
validateMongo(config.mongo.uri, spinner)
  .then(() => validateStorage(config.storage.type, config.storage.uploadDir, spinner))
  .then(() => spinner.succeed('System check complete.'))
  .catch((err) => {
    console.error(renderNativeError(err));
    spinner.fail('System check failed.');
  })
```